docs/internal/future_optimisations.md

```diff
@@ -0,0 +1,163 @@
+
+```c
+if (object->descriptor == ClassA)
+	ClassAMethod(object);
+else if (object->descriptor == ClassB)
+	ClassBMethod(object);
+```
+
+This transform is often done by dynamic language VMs because it both plays well with branch predictors and because it exposes inlining opportunities.
+It may be beneficial to inline one or both of the methods, which improves locality of the code and then exposes further optimisation opportunities from specialising the method for this specific call site.
+
+Descriptor Elision
+------------------
+
+The descriptor is used for dynamic dispatch and for `match` expressions.
+If a class is never cast to an interface type (including a union type) then it should be possible to completely elide its descriptor.
+This is particularly useful for small objects, where the descriptor pointer is a large part of the total size.
+
+Region-metadata elision
+-----------------------
+
+Objects created as `embed` fields or as entries in arrays need region metadata because they can have their address taken and be passed to other functions that are not aware of the container.
+If reachability determines that they are not captured by the called functions then there are two cases to consider:
+
+ - If the region is used then the called functions could be specialised to pass the region of the container as a separate argument.
+ - If the region is not used then the region metadata is never accessed.
+
+The second case is a clear size reduction, particularly for objects used as keys in hash tables and so on.
+The first case may result in significant code-size increases and so is worth doing only after applying a cost model.
+
+The combination of this and descriptor elision should make it possible to have efficient unit types.
+For example, consider something like this trivial example:
+
+```verona
+class Seconds
+{
+	var _value : U64;
+	+(self: Self & mut, inc : U64)
+	{
+		_value = inc;
+	}
+}
+```
+
+In the MVP compiler, this will require three words (the region metadata, the descriptor, and the 64-bit integer).
+On a CHERI system, this ends up being 40 bytes to store an 8-byte quantity.
+Now consider the following (assuming `Nanoseconds` has the same layout as `Seconds`):
+
+```verona
+class TimeInterval
+{
+	embed _seconds     : Seconds;
+	embed _nanoseconds : Nanoseconds;
+	// Methods here.
+	<(self : Self, other : TimeInterval) { ... }
+}
+```
+
+This class stores 16 bytes of state but would require 96 bytes in the MVP compiler (on a non-CHERI 64-bit system).
+Both `_seconds` and `_nanoseconds` are used by only this class and so we should be able to elide both of their descriptors and region metadata fields, shrinking this class to 32 bytes.
+If `TimeInterval` is never used in an interface or union type, then we can shrink it to 24 bytes.
+If nothing uses `TimeInterval` for region identification then we can shrink `embed` fields of type `TimeInterval` to 16 bytes.
```

Comment:
If enough inlining and static analysis is done, we can even reduce all of that to `U64` native operations ('+', '<', etc) and not have a class structure, descriptor, metadata, etc., reducing the "size" of that class' members to 8+8 bytes in total.